<a href="https://colab.research.google.com/github/mihhwiNick/DA_project/blob/huy/UK_Road_Traffic_Accident_DA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **PHÂN TÍCH VÀ DỰ BÁO XU HƯỚNG TAI NẠN GIAO THÔNG TỪ DỮ LIỆU LỚN TẠI ANH QUỐC**


## **1. Định nghĩa vấn đề (Define Problem)**

### **1.1. Mô tả (Description)**
+ Phạm vi: Dữ liệu về An toàn Giao thông Đường bộ Vương quốc Anh (UK Road Safety: Traffic Accidents and Vehicles) giai đoạn 2005-2017. Dữ liệu được cung cấp chính thức từ UK Department for Transport (Bộ Giao thông Vận tải Anh).
+ Kích thước: Gồm 2 tệp dữ liệu chính với hơn 2 triệu bản ghi về tai nạn và thông tin phương tiện liên quan.
+ Đặc điểm: Gồm 2 tệp (Accident_information và Vehicle_information) liên kết với nhau qua khóa Accident_Index. Dữ liệu có tính uy tín cao, được sử dụng rộng rãi trên Kaggle (140K lượt xem, 17.3K lượt tải).

### **1.2. Dữ liệu vào (Input Data)**
+ **A. Tệp Accident_Information (Thông tin tai nạn):**
  + **Nhóm 1: Thông tin thời gian:**
    + Date: Ngày xảy ra tai nạn.
    + Time: Thời gian xảy ra tai nạn.
    + Day_of_Week: Thứ trong tuần.
    + Year: Năm xảy ra tai nạn.
  + **Nhóm 2: Thông tin địa lí và vị trí:**
    + Latitude: Vĩ độ.
    + Longitude: Kinh độ.
    + Location_Easting_OSGR: Tọa độ đông OSGR.
    + Location_Northing_OSGR: Tọa độ bắc OSGR.
    + Urban_or_Rural_Area: Khu vực thành thị hay nông thôn.
    + Local_Authority_(District): Khu vực hành chính địa phương.
    + Local_Authority_(Highway): Cơ quan quản lý đường bộ địa phương.
    + LSOA_of_Accident_Location: Khu vực thống kê của vị trí tai nạn.
  + **Nhóm 3: Đặc điểm đường và giao lộ:**
    + 1st_Road_Class, 1st_Road_Number: Hạng và số hiệu đường chính
    + 2nd_Road_Class, 2nd_Road_Number: Hạng và số hiệu đường thứ hai
    + Road_Type: Loại đường
    + Speed_Limit: Tốc độ giới hạn
    + Junction_Control, Junction_Detail: Cách kiểm soát và chi tiết giao lộ
  + **Nhóm 4: Điều kiện môi trường và hiện trường:**
    + Light_Conditions: Điều kiện ánh sáng
    + Weather_Conditions: Điều kiện thời tiết
    + Road_Surface_Conditions: Tình trạng bề mặt đường
    + Special_Conditions_at_Site: Điều kiện đặc biệt tại hiện trường
    + Carriageway_Hazards: Nguy cơ/chướng ngại vật trên đường xe chạy
  + **Nhóm 5: Hậu quả và đáp ứng:**
    + Number_of_Vehicles: Số phương tiện liên quan
    + Number_of_Casualties: Số lượng nạn nhân
    + Did_Police_Officer_Attend_Scene_of_Accident: Cảnh sát có mặt tại hiện trường không?
    + Police_Force: Lực lượng cảnh sát ghi nhận
    + Pedestrian_Crossing-Human_Control: Người điều khiển vạch sang đường
    + Pedestrian_Crossing-Physical_Facilities: Cơ sở vật chất cho người đi bộ
    + InScotland: Tai nạn có xảy ra ở Scotland không

+ **B. Tệp Vehicle_Information (Thông tin phương tiện)**

  + **Nhóm 1: Thông tin phương tiện:**
    + Vehicle_Type: Loại xe
    + make, model: Hãng xe và mẫu xe
    + Age_of_Vehicle: Tuổi của xe
    + Engine_Capacity_.CC.: Dung tích động cơ
    + Propulsion_Code: Loại động cơ
    + Was_Vehicle_Left_Hand_Drive: Có phải xe tay lái nghịch không
    + Year: Năm sản xuất phương tiện
  + **Nhóm 2: Thông tin người lái và hành trình:**
    + Sex_of_Driver: Giới tính của tài xế
    + Age_Band_of_Driver: Nhóm tuổi của tài xế
    + Journey_Purpose_of_Driver: Mục đích chuyến đi
    + Driver_Home_Area_Type: Loại khu vực cư trú của tài xế
    + Driver_IMD_Decile: Mức độ thiếu thốn của tài xế
  + **Nhóm 3: Động lực học va chạm:**
    + Vehicle_Manoeuvre: Thao tác của xe trước khi va chạm
    + X1st_Point_of_Impact: Điểm va chạm đầu tiên
    + Vehicle_Leaving_Carriageway: Xe có rời khỏi phần đường xe chạy không
    + Skidding_and_Overturning: Có trượt bánh và lật xe không
    + Hit_Object_in_Carriageway: Vật thể bị va chạm trên làn đường
    + Hit_Object_off_Carriageway: Vật thể bị va chạm ngoài làn đường
    + Junction_Location: Vị trí tại giao lộ
    + Towing_and_Articulation: Tình trạng kéo theo hoặc đầu kéo
    + Vehicle_Location.Restricted_Lane: Vị trí xe trên làn đường bị hạn chế

### **1.3. Dữ liệu ra (Output Data)**
  + Biến mục tiêu (Output variable): **Accident_Severity**: Mức độ nghiêm trọng của tai nạn
  + Kết quả phân tích:
    + Các phát hiện/Insight về thực trạng, xu hướng và các yếu tố then chốt ảnh hưởng đến tai nạn giao thông.
    + Mô hình dự báo mức độ nghiêm trọng của tai nạn.
    + Các giải pháp và khuyến nghị chính sách.

### **1.4. Mục tiêu nghiên cứu (Research Objective)**
+ Phân tích thực trạng và xu hướng tai nạn giao thông theo thời gian, không gian và các điều kiện vận hành.
+ Xác định các yếu tố then chốt ảnh hưởng đến tần suất và mức độ nghiêm trọng của tai nạn.
+ Xây dựng và đánh giá mô hình dự báo mức độ nghiêm trọng tai nạn dựa trên các thuật toán học máy.
+ Đề xuất các giải pháp và khuyến nghị chính sách dựa trên kết quả phân tích.

##**2. Chuẩn bị vấn đề (Prepare Problem)**

###**2.1. Khai báo thư viện (Load Libraries)**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

###**2.2. Nạp dữ liệu (Load Dataset)**

In [6]:
try:
    Accident = pd.read_csv('/content/gdrive/MyDrive/Dataset_Of_DA/Accident_Information.csv',
                          encoding='latin1',
                          engine='python')
    Vehicle = pd.read_csv('/content/gdrive/MyDrive/Dataset_Of_DA/Vehicle_Information.csv',
                         encoding='latin1',
                         engine='python')
    print("Đã load dữ liệu thành công!")
except Exception as e:
    print(f"Lỗi: {e}")

Đã load dữ liệu thành công!
Kích thước Accident: (2047256, 34)
Kích thước Vehicle: (2177205, 24)


##**3. Thống kê mô tả (Descriptive Statistics)**

### **3.1. File Accident_Information**

#### **(1) Hiển thị một số thông tin về dữ liệu**
+ Số dòng, số cột của dữ liệu
+ 2 dòng đầu và 2 dòng cuối của bảng dữ liệu
+ Thông tin chung về dữ liệu

In [11]:
# shape
print(f'+ Shape: {Accident.shape}')
# head, tail
print(f'+ Contents: ')
display(Accident.head(2))
display(Accident.tail(2))
# info
Accident.info()

+ Shape: (2047256, 34)
+ Contents: 


,Accident_Index,1st_Road_Class,1st_Road_Number,2nd_Road_Class,2nd_Road_Number,Accident_Severity,Carriageway_Hazards,Date,Day_of_Week,Did_Police_Officer_Attend_Scene_of_Accident,...,Police_Force,Road_Surface_Conditions,Road_Type,Special_Conditions_at_Site,Speed_limit,Time,Urban_or_Rural_Area,Weather_Conditions,Year,InScotland
0,200501BS00001,A,3218.0,NaN,0.0,Serious,NaN,2005-01-04,Tuesday,1.0,...,Metropolitan Police,Wet or damp,Single carriageway,NaN,30.0,17:42,Urban,Raining no high winds,2005,No
1,200501BS00002,B,450.0,C,0.0,Slight,NaN,2005-01-05,Wednesday,1.0,...,Metropolitan Police,Dry,Dual carriageway,NaN,30.0,17:36,Urban,Fine no high winds,2005,No


,Accident_Index,1st_Road_Class,1st_Road_Number,2nd_Road_Class,2nd_Road_Number,Accident_Severity,Carriageway_Hazards,Date,Day_of_Week,Did_Police_Officer_Attend_Scene_of_Accident,...,Police_Force,Road_Surface_Conditions,Road_Type,Special_Conditions_at_Site,Speed_limit,Time,Urban_or_Rural_Area,Weather_Conditions,Year,InScotland
2047254,2017984122317,A,708.0,NaN,0.0,Slight,NaN,2017-07-18,Tuesday,1.0,...,Dumfries and Galloway,Dry,Single carriageway,NaN,60.0,18:00,Rural,Fine no high winds,2017,Yes
2047255,2017984122617,B,721.0,B,724.0,Serious,NaN,2017-12-20,Wednesday,1.0,...,Dumfries and Galloway,Wet or damp,Single carriageway,NaN,40.0,13:00,Rural,Fog or mist,2017,Yes


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2047256 entries, 0 to 2047255
Data columns (total 34 columns):
 #   Column                                       Dtype  
---  ------                                       -----  
 0   Accident_Index                               object 
 1   1st_Road_Class                               object 
 2   1st_Road_Number                              float64
 3   2nd_Road_Class                               object 
 4   2nd_Road_Number                              float64
 5   Accident_Severity                            object 
 6   Carriageway_Hazards                          object 
 7   Date                                         object 
 8   Day_of_Week                                  object 
 9   Did_Police_Officer_Attend_Scene_of_Accident  float64
 10  Junction_Control                             object 
 11  Junction_Detail                              object 
 12  Latitude                                     float64
 13  Light_Condit

**Thông tin cơ bản của dữ liệu:**
+ Dữ liệu có 34 tính chất (cột) để phân tích, bao gồm: Accident_Index, 1st_Road_Class, 1st_Road_Number, 2nd_Road_Class, 2nd_Road_Number, Accident_Severity, Carriageway_Hazards, Date, Day_of_Week, Did_Police_Officer_Attend_Scene_of_Accident, Junction_Control, Junction_Detail, Latitude, Light_Conditions, Local_Authority_(District), Local_Authority_(Highway), Location_Easting_OSGR, Location_Northing_OSGR, Longitude, LSOA_of_Accident_Location, Number_of_Casualties, Number_of_Vehicles, Pedestrian_Crossing-Human_Control, Pedestrian_Crossing-Physical_Facilities, Police_Force, Road_Surface_Conditions, Road_Type, Special_Conditions_at_Site, Speed_limit, Time, Urban_or_Rural_Area, Weather_Conditions, Year, và InScotland.
+ Các giá trị thuộc đa số thuộc kiểu object (chuỗi/text) với 21 cột, 10 kiểu float64 (các cột số thực như tọa độ, số đường), và 3 kiểu int64 (các cột Number_of_Casualties, Number_of_Vehicles, Year).
+ Tổng số dòng dữ liệu là 2,047,256 bản ghi (vụ tai nạn).
+ Bộ dữ liệu này không có cột phân lớp. Đây là dữ liệu an toàn giao thông/ghi chép tai nạn, trong đó cột Accident_Severity (mức độ nghiêm trọng tai nạn) là biến chính để phân tích.

#### **(2) Kiểm tra tính toàn vẹn của dữ liệu**
+ Dữ liệu có bị trùng lặp không? Hiển thị dòng bị vi phạm.
+ Dữ liệu có tồn tại giá trị Null không? Hiển thị dòng bị vi phạm.
+ Dữ liệu có tồn tại giá trị NaN không? Hiển thị dòng bị vi phạm.

In [25]:
# Kiểm tra giá trị Null hoặc NaN
has_null = Accident.isnull().sum().any()
null_count = Accident.isnull().sum().sum()
if has_null:
    print(f'- Tổng số giá trị Null: {null_count}')
    null_rows = Accident[Accident.isnull().any(axis=1)]
    print(f'- Tổng số dòng có giá trị Null hoặc NaN: {len(null_rows)}')

# Kiểm tra dữ liệu trùng lặp
n_duplicated = Accident.duplicated().sum()
print(f'- Số dòng trùng lặp: {n_duplicated}')

# Thống kê chi tiết các cột có giá trị missing
print("- Thống kê chi tiết giá trị missing theo cột:")
missing_stats = Accident.isnull().sum()
missing_percent = (Accident.isnull().sum() / len(Accident)) * 100
missing_df = pd.DataFrame({
    'Số lượng missing': missing_stats,
    'Tỷ lệ %': missing_percent
}).sort_values('Số lượng missing', ascending=False)
missing_df = missing_df[missing_df['Số lượng missing'] > 0]
display(missing_df)

- Tổng số giá trị Null: 5017445
- Tổng số dòng có giá trị Null hoặc NaN: 2045811
- Số dòng trùng lặp: 0
- Thống kê chi tiết giá trị missing theo cột:


,Số lượng missing,Tỷ lệ %
Carriageway_Hazards,2007807,98.073079
Special_Conditions_at_Site,1995137,97.454202
2nd_Road_Class,844272,41.239200
LSOA_of_Accident_Location,144953,7.080355
2nd_Road_Number,17593,0.859345
Pedestrian_Crossing-Physical_Facilities,3560,0.173891
Pedestrian_Crossing-Human_Control,2920,0.142630
Did_Police_Officer_Attend_Scene_of_Accident,278,0.013579
Longitude,175,0.008548
Latitude,174,0.008499


**Nhận xét:**
+ Dữ liệu không có dòng nào bị trùng lặp (0 dòng trùng)
+ Dữ liệu có tồn tại giá trị missing với tổng số 5,017,445 giá trị Null/NaN
+ Có 2,045,811 dòng chứa ít nhất 1 giá trị missing (chiếm ~99.9% tổng số dòng)
+ Các cột Carriageway_Hazards và Special_Conditions_at_Site có tỷ lệ missing rất cao (>97%), có thể cần xem xét loại bỏ hoặc xử lý đặc biệt
+ Các cột 2nd_Road_Class có tỷ lệ missing khá cao (41.24%)
+ Hầu hết các cột khác có tỷ lệ missing thấp (<8%) hoặc rất thấp (<1%)

#### **(3) Các tính chất thống kê trên dữ liệu số**
+ Count, Mean, Standard Deviation, Minimum Value
+ 25th Percentile, 50th Percentile (Median), 75th Percentile, Maximum Value

In [23]:
# Lọc các cột số
numeric_accident = Accident.select_dtypes(include=['number'])
description_accident = numeric_accident.describe().T

# Thêm variance (phương sai)
description_accident['variance'] = numeric_accident.var()
description_accident['datatype'] = numeric_accident.dtypes

# Hiển thị kết quả
display(description_accident[['count', 'mean', 'std', 'variance', 'min', '25%', '50%', '75%', 'max', 'datatype']])

,count,mean,std,variance,min,25%,50%,75%,max,datatype
1st_Road_Number,2047254.0,992.105099,1809.407936,3.273957e+06,0.000000,0.000000,118.000000,702.000000,9.999000e+03,float64
2nd_Road_Number,2029663.0,372.815342,1287.796383,1.658420e+06,0.000000,0.000000,0.000000,0.000000,9.999000e+03,float64
Did_Police_Officer_Attend_Scene_of_Accident,2046978.0,1.202319,0.408194,1.666220e-01,1.000000,1.000000,1.000000,1.000000,3.000000e+00,float64
Latitude,2047082.0,52.559780,1.445506,2.089488e+00,49.912941,51.485404,52.237583,53.455902,6.075754e+01,float64
Location_Easting_OSGR,2047092.0,441446.210036,95496.198208,9.119524e+09,64950.000000,378063.500000,443050.000000,524298.250000,6.555400e+05,float64
Location_Northing_OSGR,2047092.0,296885.465668,160527.257527,2.576900e+10,10290.000000,177756.750000,261183.500000,395610.000000,1.208800e+06,float64
Longitude,2047081.0,-1.410155,1.403532,1.969903e+00,-7.516225,-2.329610,-1.362233,-0.205260,1.762010e+00,float64
Number_of_Casualties,2047256.0,1.345843,0.817963,6.690630e-01,1.000000,1.000000,1.000000,1.000000,9.300000e+01,int64
Number_of_Vehicles,2047256.0,1.833525,0.715054,5.113026e-01,1.000000,1.000000,2.000000,2.000000,6.700000e+01,int64
Pedestrian_Crossing-Human_Control,2044336.0,0.010417,0.135113,1.825540e-02,0.000000,0.000000,0.000000,0.000000,2.000000e+00,float64


#### **(4) Các tính chất thống kê trên dữ liệu phân loại**
+ Count, Unique, Top, Frequency, Data Type.

In [24]:
# Lọc các cột object
object_accident = Accident.select_dtypes(include=['object'])
description_object = object_accident.describe().T

# Thêm các thông tin bổ sung
description_object['unique_count'] = object_accident.nunique()
description_object['most_frequent_count'] = object_accident.apply(lambda x: x.value_counts().iloc[0])
description_object['most_frequent_percent'] = (description_object['most_frequent_count'] / len(Accident)) * 100
description_object['datatype'] = object_accident.dtypes

# Hiển thị kết quả
display(description_object[['count', 'unique', 'top', 'freq', 'most_frequent_count', 'most_frequent_percent', 'datatype']])

,count,unique,top,freq,most_frequent_count,most_frequent_percent,datatype
Accident_Index,2047256,2047256,2017984122617,1,1,0.000049,object
1st_Road_Class,2047256,6,A,926729,926729,45.266884,object
2nd_Road_Class,1202984,6,Unclassified,814983,814983,39.808553,object
Accident_Severity,2047256,3,Slight,1734548,1734548,84.725506,object
Carriageway_Hazards,39449,6,Other object on road,16111,16111,0.786956,object
Date,2047256,4748,2005-10-21,822,822,0.040151,object
Day_of_Week,2047256,7,Friday,335183,335183,16.372305,object
Junction_Control,2047256,6,Give way or uncontrolled,988313,988313,48.275008,object
Junction_Detail,2047256,10,Not at junction or within 20 metres,827223,827223,40.406427,object
Light_Conditions,2047256,6,Daylight,1496121,1496121,73.079332,object


### **3.2. File Vehicle_Information**

#### **(1) Hiển thị một số thông tin về dữ liệu**
+ Số dòng, số cột của dữ liệu
+ 2 dòng đầu và 2 dòng cuối của bảng dữ liệu
+ Thông tin chung về dữ liệu

In [12]:
# shape
print(f'+ Shape: {Vehicle.shape}')
# head, tail
print(f'+ Contents: ')
display(Vehicle.head(2))
display(Vehicle.tail(2))
# info
Vehicle.info()

+ Shape: (2177205, 24)
+ Contents: 


,Accident_Index,Age_Band_of_Driver,Age_of_Vehicle,Driver_Home_Area_Type,Driver_IMD_Decile,Engine_Capacity_.CC.,Hit_Object_in_Carriageway,Hit_Object_off_Carriageway,Journey_Purpose_of_Driver,Junction_Location,...,Skidding_and_Overturning,Towing_and_Articulation,Vehicle_Leaving_Carriageway,Vehicle_Location.Restricted_Lane,Vehicle_Manoeuvre,Vehicle_Reference,Vehicle_Type,Was_Vehicle_Left_Hand_Drive,X1st_Point_of_Impact,Year
0,200401BS00001,26 - 35,3.0,Urban area,4.0,1588.0,NaN,NaN,Data missing or out of range,Data missing or out of range,...,NaN,No tow/articulation,Did not leave carriageway,0.0,Going ahead other,2,109,Data missing or out of range,Front,2004
1,200401BS00002,26 - 35,NaN,Urban area,3.0,NaN,NaN,NaN,Data missing or out of range,Data missing or out of range,...,NaN,No tow/articulation,Did not leave carriageway,0.0,Going ahead other,1,109,Data missing or out of range,Front,2004


,Accident_Index,Age_Band_of_Driver,Age_of_Vehicle,Driver_Home_Area_Type,Driver_IMD_Decile,Engine_Capacity_.CC.,Hit_Object_in_Carriageway,Hit_Object_off_Carriageway,Journey_Purpose_of_Driver,Junction_Location,...,Skidding_and_Overturning,Towing_and_Articulation,Vehicle_Leaving_Carriageway,Vehicle_Location.Restricted_Lane,Vehicle_Manoeuvre,Vehicle_Reference,Vehicle_Type,Was_Vehicle_Left_Hand_Drive,X1st_Point_of_Impact,Year
2177203,2016984131316,16 - 20,13.0,Urban area,6.0,1796.0,NaN,Tree,Not known,Not at or within 20 metres of junction,...,Skidded,No tow/articulation,Offside,0.0,Going ahead other,1,Car,No,Front,2016
2177204,2016984133416,46 - 55,5.0,Urban area,NaN,2184.0,NaN,Wall or fence,Not known,Not at or within 20 metres of junction,...,Skidded and overturned,No tow/articulation,Nearside,0.0,Going ahead other,1,Car,No,Nearside,2016


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2177205 entries, 0 to 2177204
Data columns (total 24 columns):
 #   Column                            Dtype  
---  ------                            -----  
 0   Accident_Index                    object 
 1   Age_Band_of_Driver                object 
 2   Age_of_Vehicle                    float64
 3   Driver_Home_Area_Type             object 
 4   Driver_IMD_Decile                 float64
 5   Engine_Capacity_.CC.              float64
 6   Hit_Object_in_Carriageway         object 
 7   Hit_Object_off_Carriageway        object 
 8   Journey_Purpose_of_Driver         object 
 9   Junction_Location                 object 
 10  make                              object 
 11  model                             object 
 12  Propulsion_Code                   object 
 13  Sex_of_Driver                     object 
 14  Skidding_and_Overturning          object 
 15  Towing_and_Articulation           object 
 16  Vehicle_Leaving_Carriageway       ob

**Thông tin cơ bản của dữ liệu:**
+ Dữ liệu có 24 tính chất (cột) để phân tích, bao gồm: Accident_Index, Age_Band_of_Driver, Age_of_Vehicle, Driver_Home_Area_Type, Driver_IMD_Decile, Engine_Capacity_.CC., Hit_Object_in_Carriageway, Hit_Object_off_Carriageway, Journey_Purpose_of_Driver, Junction_Location, make, model, Propulsion_Code, Sex_of_Driver, Skidding_and_Overturning, Towing_and_Articulation, Vehicle_Leaving_Carriageway, Vehicle_Location.Restricted_Lane, Vehicle_Manoeuvre, Vehicle_Reference, Vehicle_Type, Was_Vehicle_Left_Hand_Drive, X1st_Point_of_Impact, và Year.
+ Các giá trị thuộc đa số thuộc kiểu object (chuỗi/text) với 18 cột, 4 kiểu float64 (các cột số thực như tuổi xe, dung tích động cơ), và 2 kiểu int64 (các cột Vehicle_Reference, Year).
+ Tổng số dòng dữ liệu là 2,177,205 bản ghi (phương tiện tham gia tai nạn).
+ Bộ dữ liệu này không có cột phân lớp. Đây là dữ liệu phương tiện tham gia tai nạn, trong đó các cột Vehicle_Type (loại phương tiện), Age_Band_of_Driver (độ tuổi tài xế), và X1st_Point_of_Impact (điểm va chạm đầu tiên) là các biến chính để phân tích.

#### **(2) Kiểm tra tính toàn vẹn của dữ liệu**
+ Dữ liệu có bị trùng lặp không? Hiển thị dòng bị vi phạm.
+ Dữ liệu có tồn tại giá trị Null không? Hiển thị dòng bị vi phạm.
+ Dữ liệu có tồn tại giá trị NaN không? Hiển thị dòng bị vi phạm.

In [26]:
# Kiểm tra giá trị Null hoặc NaN
has_null = Vehicle.isnull().sum().any()
null_count = Vehicle.isnull().sum().sum()
if has_null:
    print(f'- Tổng số giá trị Null: {null_count}')
    null_rows = Vehicle[Vehicle.isnull().any(axis=1)]
    print(f'- Tổng số dòng có giá trị Null hoặc NaN: {len(null_rows)}')

# Kiểm tra dữ liệu trùng lặp
n_duplicated = Vehicle.duplicated().sum()
print(f'- Số dòng trùng lặp: {n_duplicated}')

# Thống kê chi tiết các cột có giá trị missing
print("- Thống kê chi tiết giá trị missing theo cột:")
missing_stats = Vehicle.isnull().sum()
missing_percent = (Vehicle.isnull().sum() / len(Vehicle)) * 100
missing_df = pd.DataFrame({
    'Số lượng missing': missing_stats,
    'Tỷ lệ %': missing_percent
}).sort_values('Số lượng missing', ascending=False)
missing_df = missing_df[missing_df['Số lượng missing'] > 0]
display(missing_df)

- Tổng số giá trị Null: 8017931
- Tổng số dòng có giá trị Null hoặc NaN: 2168261
- Số dòng trùng lặp: 0
- Thống kê chi tiết giá trị missing theo cột:


,Số lượng missing,Tỷ lệ %
Hit_Object_in_Carriageway,2087824,95.894691
Hit_Object_off_Carriageway,1989741,91.389695
Skidding_and_Overturning,1898208,87.185543
Driver_IMD_Decile,734812,33.750244
Age_of_Vehicle,358149,16.449944
model,325331,14.942598
Engine_Capacity_.CC.,265861,12.211115
Propulsion_Code,245843,11.291679
make,110845,5.091160
Vehicle_Location.Restricted_Lane,1317,0.060490


#### **(3) Các tính chất thống kê trên dữ liệu số**
+ Count, Mean, Standard Deviation, Minimum Value
+ 25th Percentile, 50th Percentile (Median), 75th Percentile, Maximum Value

In [27]:
# Lọc các cột số
numeric_accident = Vehicle.select_dtypes(include=['number'])
description_accident = numeric_accident.describe().T

# Thêm variance (phương sai)
description_accident['variance'] = numeric_accident.var()
description_accident['datatype'] = numeric_accident.dtypes

# Hiển thị kết quả
display(description_accident[['count', 'mean', 'std', 'variance', 'min', '25%', '50%', '75%', 'max', 'datatype']])

,count,mean,std,variance,min,25%,50%,75%,max,datatype
Age_of_Vehicle,1819056.0,7.108167,4.725886,2.233400e+01,1.0,3.0,7.0,10.0,111.0,float64
Driver_IMD_Decile,1442393.0,5.387559,2.821651,7.961712e+00,1.0,3.0,5.0,8.0,10.0,float64
Engine_Capacity_.CC.,1911344.0,2042.233961,1950.143170,3.803058e+06,1.0,1299.0,1598.0,1997.0,96000.0,float64
Vehicle_Location.Restricted_Lane,2175888.0,0.107304,0.879964,7.743373e-01,0.0,0.0,0.0,0.0,9.0,float64
Vehicle_Reference,2177205.0,1.553405,0.775248,6.010101e-01,1.0,1.0,1.0,2.0,91.0,int64
Year,2177205.0,2010.934147,3.694375,1.364841e+01,2004.0,2008.0,2011.0,2014.0,2016.0,int64


#### **(4) Các tính chất thống kê trên dữ liệu phân loại**
+ Count, Unique, Top, Frequency, Data Type.

In [28]:
# Lọc các cột object
object_accident = Vehicle.select_dtypes(include=['object'])
description_object = object_accident.describe().T

# Thêm các thông tin bổ sung
description_object['unique_count'] = object_accident.nunique()
description_object['most_frequent_count'] = object_accident.apply(lambda x: x.value_counts().iloc[0])
description_object['most_frequent_percent'] = (description_object['most_frequent_count'] / len(Accident)) * 100
description_object['datatype'] = object_accident.dtypes

# Hiển thị kết quả
display(description_object[['count', 'unique', 'top', 'freq', 'most_frequent_count', 'most_frequent_percent', 'datatype']])

,count,unique,top,freq,most_frequent_count,most_frequent_percent,datatype
Accident_Index,2177205,1488981,2013460234852,53,53,0.002589,object
Age_Band_of_Driver,2177205,12,26 - 35,450531,450531,22.006579,object
Driver_Home_Area_Type,2177205,4,Urban area,1436598,1436598,70.171879,object
Hit_Object_in_Carriageway,89381,12,Kerb,34354,34354,1.678051,object
Hit_Object_off_Carriageway,187464,12,Other permanent object,56640,56640,2.766630,object
Journey_Purpose_of_Driver,2177205,8,Not known,834626,834626,40.768033,object
Junction_Location,2177205,10,Not at or within 20 metres of junction,808108,808108,39.472738,object
make,2066360,535,VAUXHALL,239650,239650,11.705913,object
model,1851874,35723,MISSING,11877,11877,0.580142,object
Propulsion_Code,1931362,12,Petrol,1143097,1143097,55.835567,object


##**4. Tiền xử lí dữ liệu (Data Preprocessing)**

##**5. Phân tích khám phá (EDA) và trực quan hóa dữ liệu (Data Visualization)**

###**5.1.Phân tích đơn biến (Univariate Analysis)**

###**5.2.Phân tích đa biến (Multivariate Analysis)**

###**5.3.Kiểm định giả thuyết (Hypothesis Testing)**

##**6. Xây dựng mô hình (Model Building)**

### **6.1. Chuẩn bị dữ liệu**

### **6.2. Logistic Regression**

### **6.3. Random Forest (RF Model)**

### **6.4. XGBoost**

##**7. Kết luận**